### ISSUE TITLE
关于在线模型

### ISSUE THREAD

**Issue by dlwsds** [jiamingNo1/Temporal-Shift-Module#5]
你好，请问你有没有尝试训练在线的模型？
    def shift(x, n_segment, fold_div=3):
        nt, c, h, w = x.size()
        n_batch = nt // n_segment
        x = x.view(n_batch, n_segment, c, h, w)

        fold = c // fold_div

        out = torch.zeros_like(x)
        out[:, :-1, :fold] = x[:, 1:, :fold]  # shift future frames
        out[:, 1:, fold: 2 * fold] = x[:, :-1, fold: 2 * fold]  # shift past frames
        out[:, :, 2 * fold:] = x[:, :, 2 * fold:]  # not shift

        return out.view(nt, c, h, w)
如果想训练在线模型，这一部分应该怎么修改？谢谢！

**Comment by jiamingNo1** [comment]
@ChiShao ，可以试试
```
fold = c // fold_div
out= torch.zeros_like(x)
out[:, 1:, :2*fold] = x[:, :-1, :2*fold]   # shift past frames
out[:, :, 2*fold:] = x[:, :, 2*fold:]        # not shift
```

**Comment by zzwei1** [comment]
> @ChiShao ，可以试试
> 
> ```
> fold = c // fold_div
> out= torch.zeros_like(x)
> out[:, 1:, :2*fold] = x[:, :-1, :2*fold]   # shift past frames
> out[:, :, 2*fold:] = x[:, :, 2*fold:]        # not shift
> ```

你好，我想请教一下TSM的输入尺寸问题。
我有一组自己的数据来做手势识别（数据形式是这样的：每个手势有连续32帧图片），我的输入数据尺寸就是（N,C,T,H,W），但是我看TSM文章里，用2D-resnet做backbone,那么他要求的数据是就是4维的。请问我该怎么操作，才能匹配上输入尺寸呢？
我看TSM代码里，作者是（NT,C,H,W）这样一个维度，但是如果我直接对我的输入进行x.view(NT,C,H,W)，那我的标签尺寸还是原来的N,还是对不上呢。

**Comment by jiamingNo1** [comment]
@zzwei1 
https://github.com/jiamingNo1/Temporal-Shift-Module/blob/f6bad75d22c4038304cf5610462fd2fc951f0f82/ops/models.py#L136
输入数据的尺寸是（N,T,C,H,W），作者在输入到model的时候进行了view操作变为（NT,C,H,W）。

**Comment by cuixing158** [comment]
@ChiShao,@jiamingNo1 ,@zzwei1
对于一般常规情况输入维度是(N,C,T,H,W),但在TSM中有点不同，在输入进网络之前，变化过程为(N,C,T,H,W)---torch.permute--->(N,T,C,H,W)---torch.view--->(N*T,C,H,W)---2D net.forward--->(N*T,numClasses)---torch.view--->(N,T,numClasses)---torch.mean(dim=1)--->(N,numClasses),即得到N个samples的预测。
